File: Baekjoon/SegmentTree/G1_1275_커피숍2/solution.py
nums, tree = [], []


def build_tree(start, end, node): # 구간 합 트리 생성
    if start == end:
        tree[node] = nums[start]
        return tree[node]
    mid = (start + end) // 2
    tree[node] = build_tree(start, mid, node*2) + build_tree(mid+1, end, node*2+1)
    return tree[node]


def get_sum(start, end, node, left, right): # left~right 구간의 합 반환
    if end < left or start > right: return 0
    if start >= left and end <= right: return tree[node]
    mid = (start + end) // 2
    return get_sum(start, mid, node*2, left, right) + get_sum(mid+1, end, node*2+1, left, right)


def change_num(start, end, node, index, diff): # index번째 수를 diff만큼 수정
    if start > index or end < index: return
    tree[node] += diff
    if start == end: return
    mid = (start + end) // 2
    change_num(start, mid, node*2, index, diff)
    change_num(mid+1, end, node*2+1, index, diff)


def main():
    global nums, tree
    n, q = map(int, input().split())
    nums = [0]+list(map(int, input().split())) # 1 2 3 4 5
    tree = [0 for _ in range((n+1)*4)]
    build_tree(0, n, 1)
    for _ in range(q):
        x, y, a, b = map(int, input().split())
        x, y = min(x, y), max(x, y)
        print(get_sum(0, n, 1, x, y))
        diff = b - nums[a]
        nums[a] = b
        change_num(0, n, 1, a, diff)


if __name__ == "__main__":
    main()

File: Baekjoon/SegmentTree/G1_1275_커피숍2/solution.js
const input = require("fs").readFileSync('example.txt').toString().trim().split("\n");
// const input = require("fs").readFileSync('/dev/stdin').toString().trim().split("\n");

const [N, Q] = input[0].split(" ").map(Number);
let nums = input[1].split(" ").map(Number);
nums.unshift(0);
const tree = Array.from({length : (N+1)*4}, () => 0);

function buildTree(start, end, node) {
    if (start == end) return tree[node] = nums[start];
    let mid = Math.floor((start+end) / 2);
    return tree[node] = buildTree(start, mid, node*2) + buildTree(mid+1, end, node*2+1);
}

function getSum(start, end, node, left, right) {
    if (left > end || right < start) return 0;
    if (left <= start && end <= right) return tree[node];
    let mid = Math.floor((start+end)/2);
    return getSum(start, mid, node*2, left, right) + getSum(mid+1, end, node*2+1, left, right);
}

function replaceNum(start, end, node, index, diff) {
    if (index < start || index > end) return;
    tree[node] += diff;
    if (start == end) return;
    let mid = Math.floor((start+end)/2);
    replaceNum(start, mid, node*2, index, diff);
    replaceNum(mid+1, end, node*2+1, index, diff);
}

function game(x, y, a, b) {
    console.log(getSum(0, N, 1, x, y));
    let diff = b-nums[a];
    nums[a] = b;
    replaceNum(0, N, 1, a, diff);
}

buildTree(0, N, 1);

for (let i=2; i<Q+2; i++) {
    let [x, y, a, b] = input[i].split(" ").map(Number);
    if (x > y) [x, y] = [y, x];
    game(x, y, a, b);
}